### PR TITLE
refactor(x/mint): introduce distributeDeveloperRewards methods; unit tests

### PR DESCRIPTION
**Pull request by p0mvn**
Part of: https://github.com/osmosis-labs/osmosis/issues/1916

Continue refactoring `DistributeMintedCoin` to make it more testable. This change extracts a separate function for distributing developer rewards (`distributeDeveloperRewards`) and covers it with tests.

This change has no state-breaking changes and does not fix rounding issues from #1917 yet. However, it should help to isolate them to make the fix easier to do.

## Brief Changelog

- move developer rewards distribution to `distributeDeveloperRewards` and unit test it
   * remove some obsolete validation such as:
       * [account address being an empty string](https://github.com/osmosis-labs/osmosis/blob/3be2640a842394e962e323e9ad1885019e4e4229/x/mint/keeper/keeper.go#L242)
           * This is not possible due to param validation here:
https://github.com/osmosis-labs/osmosis/blob/c0052bbdaebac0ec94085e7742184ae2141706b9/x/mint/types/params.go#L226-L229
           * It has also proved to be untestable. Instead, a `CONTRACT`/invariant was made for this method:
https://github.com/osmosis-labs/osmosis/blob/c0052bbdaebac0ec94085e7742184ae2141706b9/x/mint/keeper/keeper.go#L265-L267 
   * add more validations such as:
       * developer vesting module account not having enough balance to distribute the dev rewards:
https://github.com/osmosis-labs/osmosis/blob/c0052bbdaebac0ec94085e7742184ae2141706b9/x/mint/keeper/keeper.go#L265-L267
    * the rest of the logic is the same / with comments added
- TestDistributeDeveloperRewards to cover any possible behavior and assert on all account state changes
- Create issues for the logic to be refactored further and document them in code:
  * #1916  (next step)
  * #1917
  * #2025

## Testing and Verifying

This change added tests and can be verified as follows:
- `TestDistributeDeveloperRewards`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/mint/keeper/keeper.go [hunk 1]
```diff
@@ -199,74 +208,27 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
 	proportions := params.DistributionProportions
 
 	// allocate staking incentives into fee collector account to be moved to on next begin blocker by staking module
-	stakingIncentivesCoin, err := k.distributeToModule(ctx, k.feeCollectorName, mintedCoin, proportions.Staking)
+	stakingIncentivesAmount, err := k.distributeToModule(ctx, k.feeCollectorName, mintedCoin, proportions.Staking)
 	if err != nil {
 		return err
 	}
 
 	// allocate pool allocation ratio to pool-incentives module account account
```
Comment (p0mvn) [inline x/mint/keeper/keeper.go:216]:
```suggestion
	// allocate pool allocation ratio to pool-incentives module account.
```

File: x/mint/keeper/keeper.go [hunk 2]
```diff
@@ -199,74 +208,27 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
 	proportions := params.DistributionProportions
 
 	// allocate staking incentives into fee collector account to be moved to on next begin blocker by staking module
```
Comment (p0mvn) [inline x/mint/keeper/keeper.go:210]:
```suggestion
	// allocate staking incentives into fee collector account to be moved to on next begin blocker by staking module account.
```

File: x/mint/keeper/keeper.go [hunk 3]
```diff
@@ -199,74 +208,27 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
...
-	poolIncentivesCoin, err := k.distributeToModule(ctx, poolincentivestypes.ModuleName, mintedCoin, proportions.PoolIncentives)
+	poolIncentivesAmount, err := k.distributeToModule(ctx, poolincentivestypes.ModuleName, mintedCoin, proportions.PoolIncentives)
 	if err != nil {
 		return err
 	}
 
-	devRewardCoin, err := getProportions(ctx, mintedCoin, proportions.DeveloperRewards)
+	// allocate dev rewards to respective accounts from developer vesting module account.
+	devRewardAmount, err := k.distributeDeveloperRewards(ctx, mintedCoin, proportions.DeveloperRewards, params.WeightedDeveloperRewardsReceivers)
+	// If no developer rewards receivers, we can found community pool instead.
```
Comment (p0mvn) [inline x/mint/keeper/keeper.go:224]:
```suggestion
```

File: x/mint/keeper/keeper.go [hunk 4]
```diff
@@ -277,19 +239,96 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
...
+	// Avoid over-allocating from the mint module address and have to later burn it here:
+	if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, devRewardCoins); err != nil {
+		return sdk.Int{}, err
+	}
+
+	// Take the current balance of the developer rewards pool and remove it from the supply offset
+	// We re-introduce the new supply at the end, in order to avoid any rounding discrepancies.
+	k.bankKeeper.AddSupplyOffset(ctx, totalMintedCoin.Denom, developerAccountBalance.Amount)
+
+	if len(developerRewardsReceivers) == 0 {
```
Comment (p0mvn) [inline x/mint/keeper/keeper.go:292]:
```suggestion
        // If no developer rewards receivers provided, fund the community pool from
        // the developer vesting module account.
	if len(developerRewardsReceivers) == 0 {
```

File: x/mint/keeper/keeper.go [hunk 5]
```diff
@@ -277,19 +238,98 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
...
+	devRewardCoin, err := getProportions(ctx, totalMintedCoin, developerRewardsProportion)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	developerRewardsModuleAccountAddress := k.accountKeeper.GetModuleAddress(types.DeveloperVestingModuleAcctName)
+	developerAccountBalance := k.bankKeeper.GetBalance(ctx, developerRewardsModuleAccountAddress, totalMintedCoin.Denom)
+	if developerAccountBalance.Amount.LT(devRewardCoin.Amount) {
+		return sdk.Int{}, insufficientDevVestingBalanceError{ActualBalance: developerAccountBalance.Amount, AttemptedDistribution: devRewardCoin.Amount}
+	}
```
Comment (ValarDragon) [inline x/mint/keeper/keeper.go:278]: This is a new check

File: x/mint/keeper/keeper.go [hunk 6]
```diff
@@ -277,19 +238,98 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
...
+	devRewardCoins := sdk.NewCoins(devRewardCoin)
+	// TODO: next step in https://github.com/osmosis-labs/osmosis/issues/1916
+	// Avoid over-allocating from the mint module address and have to later burn it here:
+	if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, devRewardCoins); err != nil {
+		return sdk.Int{}, err
+	}
+
+	// Take the current balance of the developer rewards pool and remove it from the supply offset
+	// We re-introduce the new supply at the end, in order to avoid any rounding discrepancies.
+	k.bankKeeper.AddSupplyOffset(ctx, totalMintedCoin.Denom, developerAccountBalance.Amount)
```
Comment (ValarDragon) [inline x/mint/keeper/keeper.go:289]: Lets consider moving this line below the community pool line, and then we don't have to adjust the community pool supply offset?

File: x/mint/keeper/keeper.go [hunk 7]
```diff
@@ -277,19 +238,98 @@ func (k Keeper) DistributeMintedCoin(ctx sdk.Context, mintedCoin sdk.Coin) error
...
+
+	// If no developer rewards receivers provided, fund the community pool from
+	// the developer vesting module account.
+	if len(developerRewardsReceivers) == 0 {
+		err = k.distrKeeper.FundCommunityPool(ctx, devRewardCoins, developerRewardsModuleAccountAddress)
+		if err != nil {
+			return sdk.Int{}, err
+		}
+		// All dev rewards go to the community pool.
+		k.bankKeeper.AddSupplyOffset(ctx, totalMintedCoin.Denom, developerAccountBalance.Amount.Sub(devRewardCoin.Amount))
```
Comment (ValarDragon) [inline x/mint/keeper/keeper.go:299]: This line is wrong / could panic

**Comment by p0mvn** [comment]
@ValarDragon thanks for the suggestion, please the first part with the direct copy of the logic here: https://github.com/osmosis-labs/osmosis/pull/2065

**Comment by p0mvn** [comment]
Closing this, will submit via 2 smaller PRs

**Pull request closed without merging**